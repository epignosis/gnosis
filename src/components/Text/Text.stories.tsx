import React from "react";
import { Story } from "@storybook/react";
import Text, { TextProps } from "./Text";

export default {
  component: Text,
  title: "Components/Text",
  argTypes: {
    weight: {
      control: {
        type: "select",
        options: ["400", "700"],
      },
    },
    fontSize: {
      control: {
        type: "select",
        // TODO: add the remaining sizes when typography docs are merged in master
        options: ["xs", "sm", "md", "lg", "xl", "2xl"],
      },
    },
    as: {
      control: {
        type: "select",
        options: ["span", "div", "p"],
      },
    },
  },
};

const Template: Story<TextProps> = (args) => <Text {...args}>This is some text!</Text>;

export const Default = Template.bind({});

Default.args = {
  weight: "400",
  fontSize: "md",
  as: "span",
};
