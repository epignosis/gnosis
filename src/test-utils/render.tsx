import React, { FC, ReactElement, ComponentType } from "react";
import { render, RenderResult, RenderOptions } from "@testing-library/react";
import ThemeProvider from "../components/ThemeProvider/ThemeProvider";

const AllTheProviders: FC = ({ children }) => <ThemeProvider>{children}</ThemeProvider>;

const customRender = (ui: ReactElement, options: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllTheProviders as ComponentType, ...options });

export * from "@testing-library/react";

export { customRender as render };
