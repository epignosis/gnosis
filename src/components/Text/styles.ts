import { css, Theme, SerializedStyles } from "@emotion/react";
import { TextWeight } from "./Text";
import { TypographyLevels } from "@theme/utils/typography";

export const textContainer = (
  { typeScaleSizes }: Theme,
  weight: TextWeight,
  size: TypographyLevels,
): SerializedStyles => css`
  font-size: ${typeScaleSizes[size]};
  font-weight: ${weight};
`;
