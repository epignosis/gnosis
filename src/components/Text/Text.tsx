import React, { FC } from "react";
import { textContainer } from "./styles";
import { TypographyLevels } from "@theme/utils/typography";

export type TextWeight = "400" | "700";

type BaseProps =
  | ({ as?: "div" } & React.HTMLAttributes<HTMLDivElement>)
  | ({ as?: "span" } & React.HTMLAttributes<HTMLSpanElement>)
  | ({ as?: "p" } & React.HTMLAttributes<HTMLParagraphElement>);

export type TextProps = BaseProps & {
  fontSize: TypographyLevels;
  weight?: TextWeight;
  className?: string;
};

const Text: FC<TextProps> = ({ children, weight = "400", fontSize, as = "span", className }) => {
  const Tag = as;

  return (
    <Tag
      data-testid="text-component"
      css={(theme) => textContainer(theme, weight, fontSize)}
      className={className}
    >
      {children}
    </Tag>
  );
};

export default Text;
