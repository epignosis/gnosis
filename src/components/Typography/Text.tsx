import React, { FC } from "react";
import { textContainer } from "./styles";
import { TypographyLevels } from "@theme/utils/typography";

export type TextWeight = "400" | "700";

export type TextProps = {
  fontSize: TypographyLevels;
  weight?: TextWeight;
  as?: "span" | "div" | "p";
  className?: string;
};

const Text: FC<TextProps> = ({ children, weight = "400", fontSize, as = "span", className }) => {
  const Tag = as;

  return (
    <Tag
      data-testid="text-component"
      css={(theme) => textContainer(theme, weight, fontSize)}
      className={className}
    >
      {children}
    </Tag>
  );
};

export default Text;
