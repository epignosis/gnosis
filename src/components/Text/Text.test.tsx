import React from "react";
import faker from "faker";
import Text from "./Text";
import { render, screen } from "@test-utils/render";

describe("<Text>", () => {
  it("renders correctly", () => {
    const text = faker.lorem.sentence(150);
    render(<Text fontSize="md">{text}</Text>);
    const textEl = screen.getByText(text);

    expect(textEl).toHaveTextContent(text);
  });

  it("snapshot", () => {
    const { container } = render(
      <Text id="my-id" className="my-class" fontSize="md">
        My content.
      </Text>,
    );

    expect(container).toMatchSnapshot();
  });
});
