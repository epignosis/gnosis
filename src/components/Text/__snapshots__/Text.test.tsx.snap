// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text> snapshot 1`] = `
<div>
  <span
    class="my-class css-1jr9658-textContainer-textContainer"
    data-testid="text-component"
    id="my-id"
  >
    My content.
  </span>
</div>
`;
